src/RevTracer.cc

```diff
@@ -0,0 +1,337 @@
+    }
+    return 0;
+    #else
+    return 1;
+    #endif
+}
+
+int SST::RevCPU::RevTracer::SetTraceSymbols(std::map<uint64_t, std::string> *TraceSymbols)
+{
+    traceSymbols = TraceSymbols;
+    return 0;
+}
+
+void SST::RevCPU::RevTracer::SetStartCycle(uint64_t c)
+{
+    startCycle = c;
+}
+
+void SST::RevCPU::RevTracer::SetCycleLimit(uint64_t c)
+{
+    cycleLimit = c;
+}
+
+void SST::RevCPU::RevTracer::SetCmdTemplate(std::string cmd)
+{
+    if (s2op.find(cmd) == s2op.end()) {
+        std::stringstream s;
+        for (auto it = s2op.begin(); it != s2op.end(); it++) {
+            s << it->first << " ";
+        }
+        pOutput->fatal(CALL_INFO, -1, "Unsupported parameter [trcCmd=%s]. Supported values are: %s\n",cmd.c_str(),s.str().c_str());
+    }
+
+    unsigned cmd_template = s2op.at(cmd);
+    for (unsigned i=0;i<NOP_COUNT;i++)
+         nops[i]= cmd_template | (i << TRC_OP_POS);
+}
+
+void SST::RevCPU::RevTracer::CheckUserControls(uint64_t cycle)
+{
+    // bail out early if disabled
+    if (disabled) return;
+    if (cycleLimit and traceCycles>cycleLimit) {
+        disabled = true;
+        if (outputEnabled) {
+            outputEnabled = false;
+            events.f.trc_ctl = 1;
+        }
+        return;
+    }
+    
+    // Using a startCycle will override programmatic controls.
+    if (startCycle>0) {
+        bool enable = startCycle and cycle > startCycle;
+        if (enable != outputEnabled) {
+            outputEnabled = enable;
+            events.f.trc_ctl = 1;
+        }
+        return;
+    }
+
+    // programatic controls
+    bool nextState = outputEnabled;
+    if (insn == nops[TRACE_OFF]) {
+        nextState = false;
+    } else if (insn == nops[TRACE_ON]) {
+        nextState = true;
+    } else if (insn == nops[TRACE_PUSH_OFF]) {
+        enableQ[enableQindex] = outputEnabled;
+        enableQindex = (enableQindex + 1 ) % MAX_ENABLE_Q;
+        nextState = false;
+    } else if (insn == nops[TRACE_PUSH_ON]) {
+        enableQ[enableQindex] = outputEnabled;
+        enableQindex = (enableQindex + 1 ) % MAX_ENABLE_Q;
+        nextState = true;
+    } else if (insn == nops[TRACE_POP]) {
+        enableQindex = (enableQindex - 1 ) % MAX_ENABLE_Q;
+        nextState = enableQ[enableQindex];
+    }
+    // prevent trace clutter with unnecessary events
+    if (nextState != outputEnabled) {
+        events.f.trc_ctl = 1;
+        outputEnabled = nextState;
+    }
+}
+
+void SST::RevCPU::RevTracer::SetFetchedInsn(uint64_t _pc, uint32_t _insn)
+{
+    insn = _insn;
+    pc = _pc;
+}
+
+bool SST::RevCPU::RevTracer::OutputOK()
+{
+    return outputEnabled or events.f.trc_ctl;
+}
+
+void SST::RevCPU::RevTracer::regRead(uint8_t r, uint64_t v)
+{
+    traceRecs.emplace_back(TraceRec_t(RegRead,r,v));
+}
+
+void SST::RevCPU::RevTracer::regWrite(uint8_t r, uint64_t v)
+{
+    traceRecs.emplace_back(TraceRec_t(RegWrite,r,v));
+}
+
+void SST::RevCPU::RevTracer::memWrite(uint64_t adr, size_t len,  const void *data)
```

Comment:
I think your namespace should be `SST::RevCPU` already so the prefix should be unecessary